In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from funpymodeling.exploratory import freq_tbl

In [4]:
df=pd.read_csv("DF_Mexico.csv")
df

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbf in position 80258: invalid start byte

**Análisis exploratorio de la base de datos Airbnb**  
1. Dimensión del data frame.
2. Tipos de datos.


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24224 entries, 0 to 24223
Data columns (total 75 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            24224 non-null  int64  
 1   listing_url                                   24224 non-null  object 
 2   scrape_id                                     24224 non-null  int64  
 3   last_scraped                                  24224 non-null  object 
 4   source                                        24224 non-null  object 
 5   name                                          24223 non-null  object 
 6   description                                   23296 non-null  object 
 7   neighborhood_overview                         15145 non-null  object 
 8   picture_url                                   24224 non-null  object 
 9   host_id                                       24224 non-null 

El DataFrame cuenta con 24,224 observaciones y 75 variables, se aprecia que contamos con tipos de datos numéricos (float e int) y con caracteres (object),
se aprecia que existen variables con datos faltantes.

**Tratamiento de datos nulos**
1. Seleccionamos las variables solicitadas por el profesor.
2. Pre- visualización de los nulos.
3. Tratamiento de datos faltantes justificando la metodología.

In [12]:
valores_nulos=df.isnull().sum()
valores_nulos

id                                                 0
listing_url                                        0
scrape_id                                          0
last_scraped                                       0
source                                             0
                                                ... 
calculated_host_listings_count                     0
calculated_host_listings_count_entire_homes        0
calculated_host_listings_count_private_rooms       0
calculated_host_listings_count_shared_rooms        0
reviews_per_month                               3596
Length: 75, dtype: int64

Dado que en la visualización no podemos observar el total de nulos pasaremos a seleccionar manualmente únicamente las variables que presenten esta condición.

In [13]:
df_seleccion = df.iloc[  : , [5,6,7,13,14,15,16,17,18,21,27,29,35,36,37,38,43,44,45,46,47,48,49,59,60,61,62,63,64,65,66,67,68,74]]
df_seleccion

name  \
0                     Amplio y luminoso loft en Coyoacán   
1                     Coyoacan Historic Studio Apartment   
2      3 Bedrooms, 2 blocks from Polanco - ALL RENOVA...   
3                                            Villa Dante   
4                                      CONDESA HAUS  B&B   
...                                                  ...   
24219                               Rincón Mexico lindo!   
24220                     Departamento en zona diamante.   
24221               Habitación privada en magnífica casa   
24222                   Espacioso y luminoso loft - Roma   
24223       Maravilloso departamento en Ciudad de México   

                                             description  \
0      Cómodo loft de dos pisos magníficamente ubicad...   
1      This studio flat is adjacent to the owner's ho...   
2      Beautifully decorated 3 bedroom apartment, it ...   
3      Dentro de Villa un estudio de arte con futon, ...   
4      A new concept of hosting in mexico through a b...   
...                                                  ...   
24219  Este lugar tiene una ubicación estratégica: ll...   
24220  Ubicado en zona diamante, junto al aeropuerto,...   
24221  Relájate en este espacio muy tranquilo y elega...   
24222  What else does it come with?<br />- WIFI 100MB...   
24223  Ven a descubrir este increíble departamento en...   

                                   neighborhood_overview        host_location  \
0                                                    NaN  Mexico City, Mexico   
1      Located in Coyoacan, in a quiet neighborhood w...  Mexico City, Mexico   
2      This area is called Nuevo Polanco, it has bein...  Mexico City, Mexico   
3      Centro comercial Santa Fe, parque interlomas y...  Mexico City, Mexico   
4                                                    NaN  Mexico City, Mexico   
...                                                  ...                  ...   
24219                                                NaN  Mexico City, Mexico   
24220                                                NaN                  NaN   
24221                                                NaN  Mexico City, Mexico   
24222                                                NaN  Mexico City, Mexico   
24223  Una zona muy bonita ya que varias de las atrac...                  NaN   

                                              host_about host_response_time  \
0                                                    NaN       within a day   
1                                                    NaN                NaN   
2      We are from CDMX and moved back and forth to N...     within an hour   
3      Master in visual arts, film photography & Mark...                NaN   
4      Condesa Haus  is Bed & Breakfast located on th...     within an hour   
...                                                  ...                ...   
24219  Amo viajar y disfrutar de mis viajes y que mej...     within an hour   
24220                                                NaN                NaN   
24221  Me considero una persona accesible para mis hu...                NaN   
24222                                                NaN     within an hour   
24223                                                NaN                NaN   

      host_response_rate host_acceptance_rate host_is_superhost  \
0                   100%                  NaN                 f   
1                    NaN                  NaN                 f   
2                   100%                  90%                 t   
3                    NaN                  NaN                 f   
4                   100%                  86%                 f   
...                  ...                  ...               ...   
24219               100%                 100%                 f   
24220                NaN                  NaN                 f   
24221                NaN                  96%                 f   
24222             

In [14]:
valores_nulos=df_seleccion.isnull().sum()
valores_nulos

name                                1
description                       928
neighborhood_overview            9079
host_location                    5172
host_about                      10664
host_response_time               3091
host_response_rate               3091
host_acceptance_rate             2043
host_is_superhost                   5
host_neighbourhood              12401
neighbourhood                    9079
neighbourhood_group_cleansed    24224
bathrooms                       24224
bathrooms_text                     30
bedrooms                          555
beds                              271
minimum_minimum_nights              1
maximum_minimum_nights              1
minimum_maximum_nights              1
maximum_maximum_nights              1
minimum_nights_avg_ntm              1
maximum_nights_avg_ntm              1
calendar_updated                24224
first_review                     3596
last_review                      3596
review_scores_rating             3596
review_score

Procedemos a eliminar las variables que no contengan datos

In [24]:
df_seleccion2 = df.iloc[  : , [5,6,7,13,14,15,16,17,18,21,27,36,37,38,43,44,45,46,47,48,59,60,61,62,63,64,65,66,67,74]]
df_seleccion2

name  \
0                     Amplio y luminoso loft en Coyoacán   
1                     Coyoacan Historic Studio Apartment   
2      3 Bedrooms, 2 blocks from Polanco - ALL RENOVA...   
3                                            Villa Dante   
4                                      CONDESA HAUS  B&B   
...                                                  ...   
24219                               Rincón Mexico lindo!   
24220                     Departamento en zona diamante.   
24221               Habitación privada en magnífica casa   
24222                   Espacioso y luminoso loft - Roma   
24223       Maravilloso departamento en Ciudad de México   

                                             description  \
0      Cómodo loft de dos pisos magníficamente ubicad...   
1      This studio flat is adjacent to the owner's ho...   
2      Beautifully decorated 3 bedroom apartment, it ...   
3      Dentro de Villa un estudio de arte con futon, ...   
4      A new concept of hosting in mexico through a b...   
...                                                  ...   
24219  Este lugar tiene una ubicación estratégica: ll...   
24220  Ubicado en zona diamante, junto al aeropuerto,...   
24221  Relájate en este espacio muy tranquilo y elega...   
24222  What else does it come with?<br />- WIFI 100MB...   
24223  Ven a descubrir este increíble departamento en...   

                                   neighborhood_overview        host_location  \
0                                                    NaN  Mexico City, Mexico   
1      Located in Coyoacan, in a quiet neighborhood w...  Mexico City, Mexico   
2      This area is called Nuevo Polanco, it has bein...  Mexico City, Mexico   
3      Centro comercial Santa Fe, parque interlomas y...  Mexico City, Mexico   
4                                                    NaN  Mexico City, Mexico   
...                                                  ...                  ...   
24219                                                NaN  Mexico City, Mexico   
24220                                                NaN                  NaN   
24221                                                NaN  Mexico City, Mexico   
24222                                                NaN  Mexico City, Mexico   
24223  Una zona muy bonita ya que varias de las atrac...                  NaN   

                                              host_about host_response_time  \
0                                                    NaN       within a day   
1                                                    NaN                NaN   
2      We are from CDMX and moved back and forth to N...     within an hour   
3      Master in visual arts, film photography & Mark...                NaN   
4      Condesa Haus  is Bed & Breakfast located on th...     within an hour   
...                                                  ...                ...   
24219  Amo viajar y disfrutar de mis viajes y que mej...     within an hour   
24220                                                NaN                NaN   
24221  Me considero una persona accesible para mis hu...                NaN   
24222                                                NaN     within an hour   
24223                                                NaN                NaN   

      host_response_rate host_acceptance_rate host_is_superhost  \
0                   100%                  NaN                 f   
1                    NaN                  NaN                 f   
2                   100%                  90%                 t   
3                    NaN                  NaN                 f   
4                   100%                  86%                 f   
...                  ...                  ...               ...   
24219               100%                 100%                 f   
24220                NaN                  NaN                 f   
24221                NaN                  96%                 f   
24222             

In [16]:
valores_nulos=df_seleccion.isnull().sum()
valores_nulos

name                               1
description                      928
neighborhood_overview           9079
host_location                   5172
host_about                     10664
host_response_time              3091
host_response_rate              3091
host_acceptance_rate            2043
host_is_superhost                  5
host_neighbourhood             12401
neighbourhood                   9079
bathrooms_text                    30
bedrooms                         555
beds                             271
minimum_minimum_nights             1
maximum_minimum_nights             1
minimum_maximum_nights             1
maximum_maximum_nights             1
minimum_nights_avg_ntm             1
maximum_nights_avg_ntm             1
first_review                    3596
last_review                     3596
review_scores_rating            3596
review_scores_accuracy          3652
review_scores_cleanliness       3651
review_scores_checkin           3652
review_scores_communication     3652
r

Vamos a separar 3 bases de datos, cada una representará cada tipo de dato.

In [25]:
df_seleccion2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24224 entries, 0 to 24223
Data columns (total 30 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   name                         24223 non-null  object 
 1   description                  23296 non-null  object 
 2   neighborhood_overview        15145 non-null  object 
 3   host_location                19052 non-null  object 
 4   host_about                   13560 non-null  object 
 5   host_response_time           21133 non-null  object 
 6   host_response_rate           21133 non-null  object 
 7   host_acceptance_rate         22181 non-null  object 
 8   host_is_superhost            24219 non-null  object 
 9   host_neighbourhood           11823 non-null  object 
 10  neighbourhood                15145 non-null  object 
 11  bathrooms_text               24194 non-null  object 
 12  bedrooms                     23669 non-null  float64
 13  beds            

**Object**

In [ ]:
object_df=df_seleccion2.iloc[  : , [0,1,2,3,4,5,6,7,8,9,10,11,20,21]]
object_df

In [37]:
#Reemplazamos datos nulos utilizando BFill
object_df2=object_df.fillna(method="bfill") 
object_df2.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24224 entries, 0 to 24223
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   name                   24224 non-null  object
 1   description            24224 non-null  object
 2   neighborhood_overview  24224 non-null  object
 3   host_location          24223 non-null  object
 4   host_about             24222 non-null  object
 5   host_response_time     24223 non-null  object
 6   host_response_rate     24223 non-null  object
 7   host_acceptance_rate   24224 non-null  object
 8   host_is_superhost      24224 non-null  object
 9   host_neighbourhood     24209 non-null  object
 10  neighbourhood          24224 non-null  object
 11  bathrooms_text         24224 non-null  object
 12  first_review           24189 non-null  object
 13  last_review            24189 non-null  object
dtypes: object(14)
memory usage: 2.6+ MB


In [38]:
#Completamos usando FFill
object_df3=object_df2.fillna(method="ffill") 
object_df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24224 entries, 0 to 24223
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   name                   24224 non-null  object
 1   description            24224 non-null  object
 2   neighborhood_overview  24224 non-null  object
 3   host_location          24224 non-null  object
 4   host_about             24224 non-null  object
 5   host_response_time     24224 non-null  object
 6   host_response_rate     24224 non-null  object
 7   host_acceptance_rate   24224 non-null  object
 8   host_is_superhost      24224 non-null  object
 9   host_neighbourhood     24224 non-null  object
 10  neighbourhood          24224 non-null  object
 11  bathrooms_text         24224 non-null  object
 12  first_review           24224 non-null  object
 13  last_review            24224 non-null  object
dtypes: object(14)
memory usage: 2.6+ MB


Validamos que los datos solicitados sean todos caracteres

In [43]:
freq_tbl(object_df3)

                                                 name  frequency  percentage  \
0                     Perfecto Loft en gran ubicación         48    0.001982   
1              Single and shared rooms near downtown.         24    0.000991   
2                              Departamento 2 cuartos         20    0.000826   
3                                 Hermosa suite doble         13    0.000537   
4      Sky Roma - Suites & Depas, Colonia Roma Norte.         12    0.000495   
...                                               ...        ...         ...   
23341             Habitación en departamento chevere!          1    0.000041   
23342       "Vive Mexico" en el corazon de Roma Norte          1    0.000041   
23343                   Renta recamara para señoritas          1    0.000041   
23344               Luxury and Spacious 3BD apartment          1    0.000041   
23345    Maravilloso departamento en Ciudad de México          1    0.000041   

       cumulative_perc  
0             

Las variables "host_response_rate" y "host_acceptance_rate" parecen ser de tipo float por lo que las enviaremos al DB correspondiente.

**Float**

In [34]:
float_df=df_seleccion.iloc[ :, [12,13,14,15,16,17,18,19,22,23,24,25,26,27,28,29]]
float_df

bedrooms  beds  minimum_minimum_nights  maximum_minimum_nights  \
0           1.0   1.0                     1.0                     1.0   
1           1.0   1.0                     3.0                     3.0   
2           3.0   3.0                    24.0                    24.0   
3           1.0   1.0                     1.0                     1.0   
4           5.0   8.0                     1.0                     1.0   
...         ...   ...                     ...                     ...   
24219       1.0   1.0                     3.0                     3.0   
24220       2.0   3.0                     1.0                     1.0   
24221       2.0   1.0                     5.0                     5.0   
24222       1.0   1.0                     1.0                     1.0   
24223       3.0   3.0                     1.0                     1.0   

       minimum_maximum_nights  maximum_maximum_nights  minimum_nights_avg_ntm  \
0                      1125.0                  1125.0                     1.0   
1                      1125.0                  1125.0                     3.0   
2                      1125.0                  1125.0                    24.0   
3                         7.0                     7.0                     1.0   
4                       365.0                   365.0                     1.0   
...                       ...                     ...                     ...   
24219                  1125.0                  1125.0                     3.0   
24220                   365.0                   365.0                     1.0   
24221                   365.0                   365.0                     5.0   
24222                   365.0                   365.0                     1.0   
24223                  1125.0                  1125.0                     1.0   

       maximum_nights_avg_ntm  review_scores_rating  review_scores_accuracy  \
0                      1125.0                  4.95                    4.71   
1                      1125.0                  4.85                    4.90   
2                      1125.0                  4.57                    4.82   
3                         7.0                   NaN                     NaN   
4                       365.0                  4.58                    4.56   
...                       ...                   ...                     ...   
24219                  1125.0                   NaN                     NaN   
24220                   365.0                   NaN                     NaN   
24221                   365.0                   NaN                     NaN   
24222                   365.0                   NaN                     NaN   
24223                  1125.0                   NaN                     NaN   

       review_scores_cleanliness  review_scores_checkin  \
0                           4.95                   5.00   
1                           4.88                   4.93   
2                           4.71                   4.92   
3                            NaN                    NaN   
4                           4.70                   4.87   
...                          ...                    ...   
24219                        NaN                    NaN   
24220                        NaN                    NaN   
24221                        NaN                    NaN   
24222                        NaN                    NaN   
24223                        NaN                    NaN   

       review_scores_communication  review_scores_location  \
0                             4.95                    5.00   
1                             5.00                    4.95   
2                             4.92                    4.58   
3                              NaN                     NaN   
4                             4.78                    4.98   
...                            ...                     ...   
24219                          NaN                     NaN   
242

In [39]:
#Reemplazamos los datos utilizando la media
float_df2=float_df.fillna(round(float_df.mean(),1))
float_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24224 entries, 0 to 24223
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   bedrooms                     24224 non-null  float64
 1   beds                         24224 non-null  float64
 2   minimum_minimum_nights       24224 non-null  float64
 3   maximum_minimum_nights       24224 non-null  float64
 4   minimum_maximum_nights       24224 non-null  float64
 5   maximum_maximum_nights       24224 non-null  float64
 6   minimum_nights_avg_ntm       24224 non-null  float64
 7   maximum_nights_avg_ntm       24224 non-null  float64
 8   review_scores_rating         24224 non-null  float64
 9   review_scores_accuracy       24224 non-null  float64
 10  review_scores_cleanliness    24224 non-null  float64
 11  review_scores_checkin        24224 non-null  float64
 12  review_scores_communication  24224 non-null  float64
 13  review_scores_lo